unittests/AST/StructuralEquivalenceTest.cpp

```diff
@@ -209,5 +209,165 @@ TEST_F(StructuralEquivalenceTest, WrongOrderOfFieldsInClass) {
   EXPECT_FALSE(testStructuralMatch(get<0>(t), get<1>(t)));
 }
 
+TEST_F(StructuralEquivalenceTest, ParamConst) {
+  auto t = makeNamedDecls("void x(int&);",
+                          "void x(const int&);", Lang_CXX, "x");
+  EXPECT_FALSE(testStructuralMatch(get<0>(t), get<1>(t)));
+}
+
+struct StructuralEquivalenceCXXMethodTest : StructuralEquivalenceTest {
+};
+
+TEST_F(StructuralEquivalenceCXXMethodTest, ReturnType) {
+  auto t = makeNamedDecls("struct foo { char x(); };",
+                          "struct foo { int x(); };", Lang_CXX);
+  EXPECT_FALSE(testStructuralMatch(get<0>(t), get<1>(t)));
+}
+
+TEST_F(StructuralEquivalenceCXXMethodTest, ReturnConst) {
+  auto t = makeNamedDecls("struct foo { char x(); };",
+                          "struct foo { const char x(); };", Lang_CXX);
+  EXPECT_FALSE(testStructuralMatch(get<0>(t), get<1>(t)));
+}
+
+TEST_F(StructuralEquivalenceCXXMethodTest, ReturnRef) {
+  auto t = makeNamedDecls("struct foo { char &x(); };",
+                          "struct foo { char &&x(); };", Lang_CXX11);
+  EXPECT_FALSE(testStructuralMatch(get<0>(t), get<1>(t)));
+}
+
+TEST_F(StructuralEquivalenceCXXMethodTest, ParamCount) {
+  auto t = makeNamedDecls("struct foo { void x(int); };",
+                          "struct foo { void x(int, int); };", Lang_CXX);
+  EXPECT_FALSE(testStructuralMatch(get<0>(t), get<1>(t)));
+}
+
+TEST_F(StructuralEquivalenceCXXMethodTest, ParamType) {
+  auto t = makeNamedDecls("struct foo { void x(int); };",
+                          "struct foo { void x(char); };", Lang_CXX);
+  EXPECT_FALSE(testStructuralMatch(get<0>(t), get<1>(t)));
+}
+
+TEST_F(StructuralEquivalenceCXXMethodTest, ParamConst) {
+  auto t = makeNamedDecls("struct foo { void x(int&); };",
+                          "struct foo { void x(const int&); };", Lang_CXX);
+  EXPECT_FALSE(testStructuralMatch(get<0>(t), get<1>(t)));
+}
+
+TEST_F(StructuralEquivalenceCXXMethodTest, ParamName) {
+  auto t = makeNamedDecls("struct foo { void x(int a); };",
+                          "struct foo { void x(int b); };", Lang_CXX);
+  EXPECT_TRUE(testStructuralMatch(get<0>(t), get<1>(t)));
+}
+
+TEST_F(StructuralEquivalenceCXXMethodTest, Virtual) {
+  auto t = makeNamedDecls("struct foo { void x(); };",
+                          "struct foo { virtual void x(); };", Lang_CXX);
+  EXPECT_FALSE(testStructuralMatch(get<0>(t), get<1>(t)));
+}
+
+TEST_F(StructuralEquivalenceCXXMethodTest, Pure) {
+  auto t = makeNamedDecls("struct foo { virtual void x(); };",
+                          "struct foo { virtual void x() = 0; };", Lang_CXX);
+  EXPECT_FALSE(testStructuralMatch(get<0>(t), get<1>(t)));
+}
+
+TEST_F(StructuralEquivalenceCXXMethodTest, Const) {
+  auto t = makeNamedDecls("struct foo { void x(); };",
+                          "struct foo { void x() const; };", Lang_CXX);
```

Comment:
Could you please add tests for the `&&` qualifier too?
```
struct foo { void x() &&; };
```
Also, for `noexcept` and `throw()` functions?
```
void f0;
void f1() throw();
void f2() noexcept;
```
f0 should be structurally eq with f1, but not with f2, because noexcept is part of the type system.

Reply:
Probably we will have to add some tests in a disabled form (`noexcept` will fail).

Reply:
So there should be no check for exception specification in "throw" form at all?
But the `noexcept` should be checked, and then check the related expressions (if exist) too?

There is check for exceptions in `isStructurallyEquivalent` (QualType) that never checks anything because the "canonical type" does not contain exception information. (It is possible to use the original non-canonical type here.)

Reply:
`throw()` should be equal to `noexcept(true)` and with `noexcept`. Any other forms of throw we should ignore.

```
void f1() noexcept(expr1);
void f2() noexcept(expr2);
```
Above `f1` and `f2` should have the same type only if both expressions evaluates to the same value. But probably this would require constexpr evaluation which we might not be able to do at this phase.

Okay, I see now it is getting complicated... perhaps this should be kept as future work.